sdk/core/Azure.Core/src/Shared/GeoRedundantReadFallbackPolicy.cs

```diff
@@ -0,0 +1,100 @@
+    {
+        private readonly string[] _readFallbackHosts;
+        private int? _fallbackIndex;
+        private readonly object _syncLock = new();
+
+        /// <summary>
+        ///
+        /// </summary>
+        /// <param name="readFallbackHosts"></param>
+        public GeoRedundantReadFallbackPolicy(string[] readFallbackHosts)
+        {
+            _readFallbackHosts = readFallbackHosts;
+        }
+
+        public static void SetHostAffinity(HttpMessage message, bool hostAffinity)
+        {
+            message.SetInternalProperty(typeof(GeoRedundantReadFallbackPolicy), hostAffinity);
+        }
+
+        public static bool GetHostAffinity(HttpMessage message)
+        {
+            return message.TryGetInternalProperty(typeof(GeoRedundantReadFallbackPolicy), out object? hostAffinity) && hostAffinity is true;
+        }
+
+        /// <summary>
+        ///
+        /// </summary>
+        /// <param name="message"></param>
+        public override void OnSendingRequest(HttpMessage message)
+        {
+            // TODO implement policy for write hosts as well (or include in this policy)
+            if (message.Request.Method != RequestMethod.Get && message.Request.Method != RequestMethod.Head)
+                return;
+
+            // must be under lock so that this can be synchronized across client calls - each call should leverage the fallback information
+            // from all other calls
+            lock (_syncLock)
+            {
+                // cannot change the host - client is responsible for setting this on a message that cannot fallback. Additionally,
+                // client must set the host to the correct host based on previous calls.
+                // Example - Method Foo makes two requests that must go to the same endpoint:
+                // var firstMessage = _pipeline.CreateMessage();
+                // ...
+                // _pipeline.Send(firstMessage);
+                //
+                // var secondMessage = _pipeline.CreateMessage();
+                // secondMessage.Request.Uri.Host = firstMessage.Request.Uri.Host;
+                // GeoRedundantReadFallbackPolicy.SetHostAffinity(secondMessage, true);
+                // secondMessage.ProcessingContext.HostAffinity = true;
+                // ...
+                // _pipeline.Send(secondMessage);
+
+                if (GetHostAffinity(message))
+                    return;
+
+                // first attempt - use fallback host if set
+                if (message.ProcessingContext.RetryNumber == 0)
+                {
+                    UpdateHost(message);
+                    return;
+                }
+
+                // subsequent attempt
+
+                // only fallback if no response was received
+                if (message.HasResponse)
+                    return;
+
+                // advance the index
+                _fallbackIndex ??= -1;
+                _fallbackIndex++;
```

Comment:
I think we need to check that the current host matches the one pointed to by `_fallbackIndex` before incrementing. Consider the following scenario:

Thread 1: 
 - traverse up policy stack to the transport
 - send request to primary and fail

Thread 2: 
 - traverse up policy stack to the transport
 - send request to primary and fail

Thread 1:
- Unroll the policy stack to `GeoRedundantReadFallbackPolicy`
- acquire lock 
- increment `_fallbackIndex` (points to secondary)
- release lock
- request send to secondary

Thread 2:
- Unroll the policy stack to `GeoRedundantReadFallbackPolicy`
- acquire lock 
- increment `_fallbackIndex` (points to primary, since it's already pointed at the last index)
- release lock
- request sent to primary again